### ISSUE TITLE
[v4] Overriding controller doesn't work in GraphQL

### ISSUE THREAD

**Issue by teinett** [strapi/strapi#12256]
### Describe the bug

I develop a blog with posts in Russian language.
It's nice to have correct looking quote marks, so I am using npm package Typograf to override quote marks in posts's controller.

Overriding controller works in REST API, but doesn't work in GraphQL.

It worked nice in Strapi v3, and recently in Strapi v4, but it doesn't work now.

### Steps to reproduce the behavior

1. Go to src/api/post/controllers/posts.js
2. Add code to change quote marks (the code is below)
3. Go to GraphQL and request the one post or posts
4. See error: the quote marks are not changed.
5. Go to REST API and request the one post or posts
6. There is no error: quote marks are replaced.

### Expected behavior

The quote marks should be replaced from `"` to `«` and `»`. 
It's the way to understand that Typograf works. 

### Screenshots

GraphQL: the error

![2022-01-20 13-43-53](https://user-images.githubusercontent.com/1038168/150325263-108dc9f7-e5ae-41a5-815c-8ed6e8d7df98.jpg)

REST API: no error

<img width="1395" alt="2022-01-20 13-50-35" src="https://user-images.githubusercontent.com/1038168/150325307-358e27bd-e424-45bc-bb5e-b989b800fabe.png">

The NextJS app with the request from GraphQL: the error

![2022-01-20 13-52-34](https://user-images.githubusercontent.com/1038168/150325338-277f72ab-17c7-44e8-b565-b4711739104a.jpg)

### Code snippets

My code of the post's controller:

```
'use strict';
// const { sanitizeEntity } = require('strapi-utils');
const Typograf = require('typograf');


/**
 *  post controller
 */

const { createCoreController } = require('@strapi/strapi').factories;

const tp = new Typograf({locale: ['ru', 'en-US']});

// module.exports = createCoreController('api::post.post');

module.exports = createCoreController('api::post.post', ({ strapi }) =>  ({

  async find(ctx) {
    // some custom logic here

    // Calling the default core action
    const { data, meta } = await super.find(ctx);

    // I don't know how to do the same as in findOne part, but it should be here

    return { data, meta };
  },

  // Method 3: Replacing a core action
  async findOne(ctx) {
    const { id } = ctx.params;
    const { query } = ctx;

    const entity = await strapi.service('api::post.post').findOne(id, query);
    // console.log(entity);

    if (entity) {
      if (entity.title) {entity.title = tp.execute(entity.title);}
      if (entity.subtitle) {entity.subtitle = tp.execute(entity.subtitle);}
      if (entity.description) {entity.description = tp.execute(entity.description);}
      if (entity.content) {entity.content = tp.execute(entity.content);}
    }
    const sanitizedEntity = await this.sanitizeOutput(entity, ctx);

    return this.transformResponse(sanitizedEntity);
  }

}));
```

### System

- Node.js version: 14.17.5
- NPM version: 6.14.14
- Strapi version: 4.0.5
- Database: SQLite
- Operating system: MacOS

### Additional context

Overriding controller worked at 4.0.2 and probably 4.0.3, but it doesn't work at 4.0.4 and 4.0.5.

**Comment by kibblerz** [comment]
From my understanding, GraphQL doesn't typically utilize the controllers.

 If you want to override the service you can, but if you want to do anything you'd typically do on the controller with graphql, you'd need to override the NexusJs resolver in src/index.js.

**Comment by kibblerz** [comment]
> > From my understanding, GraphQL doesn't typically utilize the controllers.
> > If you want to override the service you can, but if you want to do anything you'd typically do on the controller with graphql, you'd need to override the NexusJs resolver in src/index.js.
> 
> Previously it worked in GraphQL too: in Strapi v3 and v4.0.2 minimum. So I think it's a bug.

When GraphQL passes in requests though, it doesn't typically pass it in the format that default controllers require. Resolution is done via services in the graphql resolvers, which have a completely different structure to the requests than the default controllers typically handle. It makes little sense to me to have a resolver calling a controller as opposed to a service.

https://github.com/strapi/strapi/blob/45020eee065af8917011cc43398f95c4c624d6ad/packages/plugins/graphql/server/services/builders/resolvers/query.js

As specified here, it looks as though the graphql queries resolve with the entityService by default. So adjusting the controller should have no effect on it.

**Comment by dlujan** [comment]
This issue has been throwing me off all day. @teinett did you find a solution. I'm thinking of just going back to REST.

**Comment by kibblerz** [comment]
@dlujan 
https://docs.strapi.io/developer-docs/latest/plugins/graphql.html#extending-the-schema
You can override the query with a custom graphql resolver set up (I use the Nexus config to override it).

Just create a query with the same name, and in the resolver use the entityService to execute it. There's services in the graphql plugin to format the results for GraphQL. I believe they were named along the lines of formatEntityQuery. Though these may not be necessary.

Another option instead of overriding the whole resolver, you can use nexusjs to override the specific type (Which would be less work and preferable).

like this:
`
const { service: getGraphQLService } = strapi.plugin('graphql')
const extensionService = getGraphQLService('extension')
extensionService.use(({ nexus, typeRegistry }) => ({
nexus.objectType({
              name: `TypeNameToOverride`,
              definition: t => {
                t.field('fieldOne', {
                  type: 'String',
                  async resolve (root, args, ctx) {
                    //perform manipulation here
                    return root
                  }
                })
              t.string('fieldTwo')
              }
            }),
}))
  `
  
  you'll need to copy the schema from what strapi generates, you can reference the nexus docs here 
  https://nexusjs.org/docs/api
  hope this helps

**Comment by dlujan** [comment]
Appreciate it!

**Comment by dlujan** [comment]
hey @kibblerz no luck so far on my end with your suggested fix. Could you tell me where you are putting that code you pasted? I've tried in the register function in src/index.js but that doesn't seem to be the right place. Thanks for your help.

**Comment by derrickmehaffy** [comment]
After further review a previous response is correct, GraphQL doesn't utilize the controller system like it did in v4. REST Controllers and GraphQL resolvers are separated.

Marking as closed

**Comment by teinett** [comment]
OK, I added this as a request in feedback website - https://feedback.strapi.io/developer-experience/p/overriding-controller-in-graphql-additional-to-rest-api

**Comment by benderillo** [comment]
> @dlujan https://docs.strapi.io/developer-docs/latest/plugins/graphql.html#extending-the-schema You can override the query with a custom graphql resolver set up (I use the Nexus config to override it).
> 
> Just create a query with the same name, and in the resolver use the entityService to execute it. There's services in the graphql plugin to format the results for GraphQL. I believe they were named along the lines of formatEntityQuery. Though these may not be necessary.
> 
> Another option instead of overriding the whole resolver, you can use nexusjs to override the specific type (Which would be less work and preferable).
> 
> like this: `const { service: getGraphQLService } = strapi.plugin('graphql') const extensionService = getGraphQLService('extension') extensionService.use(({ nexus, typeRegistry }) => ({ nexus.objectType({ name:`TypeNameToOverride`, definition: t => { t.field('fieldOne', { type: 'String', async resolve (root, args, ctx) { //perform manipulation here return root } }) t.string('fieldTwo') } }), })) `
> 
> you'll need to copy the schema from what strapi generates, you can reference the nexus docs here https://nexusjs.org/docs/api hope this helps

This is a downgrade of developer experience for whoever works with GraphQL. This approach requires significantly more unneeded boilerplate to get things done.

If GraphQL resolvers (shadowCRUD) used the core services provided by corresponding APIs that would make for a better experience.
In v3. they used the controllers (shared with REST API) which is no longer the case in v4.
As far as I know, graphql resolvers directly use `strapy.entityService` which makes it impossible to alter behaviour of any of the generated resolver without registering an extension that essentially overwrites it.

I think if generated resolvers change to use `strapi.service("api::restaurant.restaurant").blah()` instead of entityService directly, it would give us, developers back that convenience we once had in v3.

**Comment by jasperem** [comment]
Please open this issue again. We would like to use Strapi with Rest and Graphql in parallel and at the moment this is only possible if we write a large part of our backend logic twice.

**Comment by GraemeFulton** [comment]
her

> This issue has been mentioned on **Strapi Community Forum**. There might be relevant details there:
> 
> https://forum.strapi.io/t/how-to-filter-graphql-results-with-controllers/47782/2

https://forum.strapi.io/t/are-graphql-queries-supposed-to-trigger-controller-actions-in-strapi-v4/13474/14?u=graeme

Here is my solution:
```
const extensionService = strapi.plugin("graphql").service("extension");

  extensionService.use({
    resolversConfig: {
      "Query.posts": {
        middlewares: [
          async (resolve, parent, args, context, info) => {
            //if user is not authenticated
            if (!context.state.user) {
              /**
               * Update the selection set
               */
              // If user is not authenticated, remove sensitive fields
              info.fieldNodes[0].selectionSet.selections.forEach((field) => {
                if (field.name.value === "specialContent") {
                  field.selectionSet.selections =
                    field.selectionSet.selections.filter(
                      (subField) => subField.name.value !== "premium",
                    );
                }
              });

              /**
               * Update the graphql filters
               */
              // Force filter free: true when user is not authenticated
              args.filters = args.filters || {};
              args.filters.free = { eq: true };

              console.log(args)
            }

            return resolve(parent, args, context, info);
          },
        ],
      },
```
